edit-post/store/effects.js

```diff
@@ -90,6 +89,31 @@ const effects = {
 		const message = action.mode === 'visual' ? __( 'Visual editor selected' ) : __( 'Code editor selected' );
 		speak( message, 'assertive' );
 	},
+	INIT( _, store ) {
```

Comment:
The new `$( document ).ready` ? This very much concerns me, as it is not at all scalable.

We seem to be contradicting ourselves that stores ought be independent from one another. What we're doing here is effectively updating a store in response to actions dispatched in other stores.

It would be as if the sidebar reducer were to include `core/editor/SELECT_BLOCK` as a switch case. This is one option. Obviously means exposing internals of editor store to other modules and its various related actions.

Another option is some API for forwarding specific action dispatchers from other stores (`forward( 'core/editor ).selectBlock( /* ... ? */ )`), though I think this suffers similarly, and either means exposing same internals, or imposing on listener to handle arguments independently.

While I don't think it applies here because it is an _effect_, not really a composed value, for many cases where we'd be inclined to add here (and it will be oh so convenient to do so), a better option would be a composing selector where a module's own selectors also select from other modules. We don't have any instances of this so far, and calling `select` within a selector feels somehow wrong, though the general idea of it seems valid.

Ultimately I think it should be something like what's done here with subscribing to selector values, though ideally in a more declarative fashion. Roughly I see three critical pieces of information:

- The selector(s) returning the value
- The condition on which a respondent action should dispatch
   - Minimally, just the new value itself
- The action to dispatch

Further, I don't even know that this needs to be a Redux-specific concept. Really it's a general pattern for responding to changes in values. I hesitate to mention it, but this is something decently covered in the [proposed Observable specification](https://github.com/tc39/proposal-observable).

Scribbling a bit:

```js
const { hasSelectedBlock } = select( 'core/editor' );
const { isViewportMatch } = select( 'core/viewport' );

new DispatchObservable( subscribe, [
	[ hasSelectedBlock, true, () => setActiveSidebar( 'edit-post/block' ) ],
	[ hasSelectedBlock, false, () => setActiveSidebar( 'edit-post/document' ) ],
	[ () => isViewportMatch( '< medium' ), true, closeGeneralSidebar ],
] ).subscribe( ( action ) => store.dispatch( action ) );
```

Where `subscribe` could be any function for starting to listen, returning an observable whose callback is invoked only when actions are to be dispatched. I think the `selectionStart && isSidebarOpened` logic could be simplified to handle the active sidebar regardless of whether or not it is currently open.

This comment is very much a brain-dump of unfinished thoughts / suggestions, but I want this to be much more unintuitive than it is as proposed, as otherwise developers _will_ add to this `INIT` effect carelessly.

Currently this gives me vibes of jQuery soup: http://tutorials.jenkov.com/jquery/critique.html#jquery-big-main-method

Reply:
Good thoughts and I agree with the general feeling that this `INIT` action is not great. And yes, the implementation here is just an implementation of returning `Observables` from other `Observables` (the original subscribe). 

You'd need to create this observable somewhere though, and that's the `INIT` action here. Previously, it was run directly after the store creation in `store/index`, this PR adds the `INIT` function as a synthetic change to gather it with the effects. But the same could be done by just moving this to a separate file and call it from the `store/index`.

Now, whether to introduce explicit `Observable`s instead of custom `onChangeListener` which is basically the exact same thing written differently, I'm open to change here.

Reply:
> We seem to be contradicting ourselves that stores ought be independent from one another. What we're doing here is effectively updating a store in response to actions dispatched in other stores.

Yes, this is the biggest concern I see with this approach. On the other hand, if you look at how modules depend on each other, it makes sense to be able to provide this kind of communication between stores. `edit-post` depends on both `editor` and `viewport` so we don't violate the overall architecture of the application.

Reply:
I don't disagree that we need to initialize some subscribing behavior. My issue is that there is no pattern for organization aligning to what it is we're expressing with these subscribe callback. So instead various behaviors will be added here, whether or not appropriate to do so, and the effect handler will grow into an unmaintainable mess.